File: src/main/java/gg/climb/lolobjects/game/ChampionStats.java
package gg.climb.lolobjects.game;

import org.immutables.value.Value;

/**
 * Champion stats as recorded by riot. See {@link ChampionStatsBuilder}
 */
@Value.Style(visibility = Value.Style.ImplementationVisibility.PRIVATE,
             strictBuilder = true)
@Value.Immutable
public abstract class ChampionStats {
  public abstract double attackRange();

  public abstract double moveSpeed();

  public abstract double attackSpeedOffset();

  public abstract double attackSpeedPerLevel();

  public abstract double attackDamage();

  public abstract double attackDamagePerLevel();

  public abstract double hp();

  public abstract double hpPerLevel();

  public abstract double hpRegen();

  public abstract double hpRegenPerLevel();

  public abstract double mp();

  public abstract double mpPerLevel();

  public abstract double mpRegen();

  public abstract double mpRegenPerLevel();

  public abstract double armor();

  public abstract double armorPerLevel();

  public abstract double spellBlock();

  public abstract double spellBlockPerLevel();

  public abstract double crit();

  public abstract double critPerLevel();

}

File: app/gg/climb/lolobjects/game/ChampionStats.scala
package gg.climb.lolobjects.game

/**
 * Champion stats as recorded by riot.
 */
class ChampionStats(attackRange: Double,
                    moveSpeed: Double,
                    attackSpeedOffset: Double,
                    attackSpeedPerLevel: Double,
                    attackDamage: Double,
                    attackDamagePerLevel: Double,
                    hp: Double,
                    hpPerLevel: Double,
                    hpRegen: Double,
                    hpRegenPerLevel: Double,
                    mp: Double,
                    mpPerLevel: Double,
                    mpRegen: Double,
                    mpRegenPerLevel: Double,
                    armor: Double,
                    armorPerLevel: Double,
                    spellBlock: Double,
                    spellBlockPerLevel: Double,
                    crit: Double,
                    critPerLevel: Double){

  override def toString = s"ChampionStats(attackRange=$attackRange, moveSpeed=$moveSpeed," +
    s"attackSpeedOffset=$attackSpeedOffset, attackSpeedPerLevel=$attackSpeedPerLevel, attackDamage=$attackDamage, " +
    s"attackDamagePerLevel=$attackDamagePerLevel, hp=$hp, hpPerLevel=$hpPerLevel, hpRegen=$hpRegen, " +
    s"hpRegenPerLevel=$hpRegenPerLevel, mp=$mp, mpPerLevel=$mpPerLevel, mpRegen=$mpRegen, " +
    s"mpRegenPerLevel=$mpRegenPerLevel, armor=$armor, armorPerLevel=$armorPerLevel, spellBlock=$spellBlock, " +
    s"spellBlockPerLevel=$spellBlockPerLevel, crit=$crit, critPerLevel=$critPerLevel)"
}
